### PR TITLE
Disallow native ropes from being used in cached tables.

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

**Comment by eregon** [comment]
Should NativeRope#hashCode() just throw or that would be too aggressive?

**Comment by nirvdrum** [comment]
I don't have a problem with that per se. But I'll need to work out why we have a custom implementation for `NativeRope#hashCode` currently.

File: src/main/java/org/truffleruby/core/symbol/SymbolTable.java [hunk 1]
```diff
@@ -96,6 +97,10 @@ public DynamicObject getSymbol(String string) {
 
     @TruffleBoundary
     public DynamicObject getSymbol(Rope rope) {
+        if (rope instanceof NativeRope) {
+            rope = ((NativeRope) rope).toLeafRope();
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/symbol/SymbolTable.java]: What's the issue to lookup a Symbol with a NativeRope?
The Rope used in the symbolMap is the one returned by RopeCache.getRope which is built from a managed `byte[]`.

**Comment by eregon** [comment]
> But I'll need to work out why we have a custom implementation for `NativeRope#hashCode` currently.

So that the hashCode is updated when the underlying bytes change (Rope#hashCode caches).

File: src/main/java/org/truffleruby/core/symbol/SymbolTable.java (hunk 1 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/symbol/SymbolTable.java]: The immediate problem is `NativeRope#withEncoding` throws an exception, which affects the `if` statement following this one. I can fix that, but I think it's better to fix the issue here.

But I think we may have another issue where the `ropeKey` stored in the symbol uses the original rope, not the one built with a managed `byte[]`.

**Comment by nirvdrum** [comment]
> So that the hashCode is updated when the underlying bytes change (Rope#hashCode caches).

I know why it works that way. I don't know what use case caused it to be added :-)

**Comment by nirvdrum** [comment]
Answering my own question, we use the rope's hash code as a component of the Ruby `String#hash`. Disallowing `NativeRope#hashCode` would require us to convert to `ManagedRope` anywhere `String#hash` is called. Maybe that's what we want. The first case I found was a native rope being used inside a `Hash`.

**Pull request merged**